File: python/LinkedList/insert_at_a_given_index.py
#!/usr/bin/python3

"""
Given the head of a LL, an index, and a value. Write a function that insert a given
node in the  LL at a given index and return the head of the List.
"""
from Node.Node import Node


def insert_at_given_index_recursively(head: Node, value: int, index: int) -> Node or None:
    """
    Recursively insert a new node at a given index in the single LL
    :param head: current head of the LL
    :param value: value to insert
    :param index: index at which to insert the value
    :return: updated head of the LL

    """

    if head is None and index == 0:
        # insert at the beginning of the list
        new_node = Node(value)
        new_node.next = head
        head = new_node
        return head

    if index == 0:
        new_node = Node(value)
        new_node.next = head
        return new_node

    head.next = insert_at_given_index_recursively(head.next, value, index - 1)
    return head


def insert_at_given_index(head: Node, value: int, index: int) -> Node:
    """
    Inserts a node in a given LL at a particular index.

    :param head: current head of the LL
    :param value: value to insert
    :param index: index at which to insert the value
    :return: updated head of the LL
    """

    if index == 0:
        # Insert at the beginning of the LL
        new_node = Node(value)
        new_node.next = head
        head = new_node
        return head

    prev_node = None
    counter = 0
    current_node = head

    while current_node is not None and counter < index:
        # Traverse the LL to find the insertion point or Not
        prev_node = current_node
        current_node = current_node.next
        counter += 1

    if counter == index:
        # Insertion point is found
        new_node = Node(value)
        prev_node.next = new_node
        new_node.next = current_node

    else:
        # index is out of bound
        print('Index is out of bound, cannot insert a new node')
        exit()

    return head

File: PHP/LinkedList/insert_at_a_given_index.php
<?php

/*
 ** Given the head of a LL, an index, and a value. Write a function that insert a given
 ** node in the  LL at a given index and return the head of the List.
 */


use LinkedList\NodeDefinition\Node;


/**
 * @param Node $head : Head of the LL
 * @param int $index : Index to insert the new node
 * @param int $value : New node's value
 * @return Node: The Head of the LL
 */
function insertNodeAtIndexIteratively(Node $head, int $index, int $value): Node
{
    if ($index === 0) {
        # insert at the beginning of the LL
        $newNode = new Node($value);
        $newNode->next = $head;
        return $newNode;
    }

    $currentNode = $head;
    $counter = 0;

    while ($currentNode !== null && $counter < $index - 1) {
        # Traverse through the LL, get to the node before insertion point
        # OR index is out of bound.
        $currentNode = $currentNode->next;
        $counter++;
    }

    if ($currentNode === null) {
        echo "Index out of bound, cannot insert a new node";
        exit();
    }

    # We find the insertion point
    $newNode = new Node($value);
    $newNode->next = $currentNode->next;
    $currentNode->next = $newNode;

    return $head;
}


/**
 * @param Node|null $head : Head of the LL
 * @param int $value : New node's value
 * @param int $index : Index to insert the new node
 * @return Node
 */

function insertNodeAtIndexRecursively(?Node $head, int $value, int $index): Node
{

    if ($index === 0) {
        $newNode = new Node($value);
        $newNode->next = $head;
        return $newNode;
    }

    $head->next = insertNodeAtIndexRecursively($head->next, $value, $index - 1);
    return $head;
}